File: shared/src/commonMain/kotlin/com/schedulo/shared/model/Feedback.kt
package com.schedulo.shared.model

/**
 * A bug report or suggestion submitted from the in-app feedback screen.
 *
 * Submissions are immutable once written: the security rules deny update and
 * delete, so a report can't change out from under whoever is triaging it. That
 * also means the screenshot has to be uploaded before the document is created —
 * there is no second write to attach the URL afterwards.
 */
data class Feedback(
    var id: String = "",
    var userId: String = "",
    var userEmail: String = "",
    var category: String = FeedbackCategory.BUG,
    var description: String = "",
    var stepsToReproduce: String = "",
    var screenshotUrl: String = "",
    var appVersion: String = "",
    var platform: String = "",
    var osVersion: String = "",
    var deviceModel: String = "",
    var status: String = "new",
    var createdAt: Long = 0
)

/**
 * The categories a report can be filed under.
 *
 * Held as plain strings rather than an enum to match the other Firestore-backed
 * models ([TeamTask.status], [SwapRequest.status]), and because the security
 * rules compare against these exact literals — an enum would put the canonical
 * spelling one conversion away from the value that actually gets validated.
 */
object FeedbackCategory {
    const val BUG = "bug"
    const val FEATURE = "feature"
    const val OTHER = "other"

    val ALL = listOf(BUG, FEATURE, OTHER)

    fun isValid(category: String): Boolean = category in ALL
}

/**
 * Length limits for the free-text fields of a report.
 *
 * These are duplicated in `firestore.rules` and mirrored in the iOS Swift code,
 * which is exactly why they live here: a client that lets the user type past the
 * server's limit turns a submission into a rules rejection at the last step,
 * after the screenshot has already been uploaded.
 */
object FeedbackLimits {
    const val MAX_DESCRIPTION = 2000
    const val MAX_STEPS = 2000

    /** A report needs something to act on, so a blank description is rejected. */
    fun isValidDescription(description: String): Boolean {
        val trimmed = description.trim()
        return trimmed.isNotEmpty() && trimmed.length <= MAX_DESCRIPTION
    }

    fun isValidSteps(steps: String): Boolean = steps.trim().length <= MAX_STEPS
}

File: iosApp/Schedulo2/Models/Feedback.swift
import Foundation

/// Where bug reports are emailed. Every report is also stored in Firestore, so a
/// change here only affects the mail hand-off, not the record of the submission.
let developerFeedbackEmail = "sheshank3336@gmail.com"

/// Mirrors `FeedbackCategory` / `FeedbackLimits` in the KMP shared module. The
/// caps have to match `firestore.rules` exactly — a client that lets the user
/// type past the server's limit turns a submission into a rules rejection at the
/// last step, after the screenshot has already been uploaded.
enum FeedbackCategory: String, CaseIterable {
    case bug
    case feature
    case other

    var label: String {
        switch self {
        case .bug: return "Bug"
        case .feature: return "Feature request"
        case .other: return "Other"
        }
    }
}

enum FeedbackLimits {
    static let maxDescription = 2000
    static let maxSteps = 2000

    /// A report needs something to act on, so a blank description is rejected.
    static func isValidDescription(_ description: String) -> Bool {
        let trimmed = description.trimmingCharacters(in: .whitespacesAndNewlines)
        return !trimmed.isEmpty && trimmed.count <= maxDescription
    }

    static func isValidSteps(_ steps: String) -> Bool {
        steps.trimmingCharacters(in: .whitespacesAndNewlines).count <= maxSteps
    }
}

struct FeedbackReport {
    var id: String = UUID().uuidString
    var userId: String = ""
    var userEmail: String = ""
    var category: String = FeedbackCategory.bug.rawValue
    var description: String = ""
    var stepsToReproduce: String = ""
    var screenshotUrl: String = ""
    var appVersion: String = ""
    var platform: String = "ios"
    var osVersion: String = ""
    var deviceModel: String = ""
    var status: String = "new"
    var createdAt: Int64 = 0

    var categoryLabel: String {
        FeedbackCategory(rawValue: category)?.label ?? "Other"
    }

    /// The plain-text report that goes into the developer's inbox.
    var emailBody: String {
        var lines: [String] = [description, ""]
        if !stepsToReproduce.isEmpty {
            lines.append("Steps to reproduce:")
            lines.append(stepsToReproduce)
            lines.append("")
        }
        if !screenshotUrl.isEmpty {
            lines.append("Screenshot: \(screenshotUrl)")
            lines.append("")
        }
        lines.append("---")
        lines.append("Category: \(categoryLabel)")
        lines.append("App version: \(appVersion)")
        lines.append("Device: \(deviceModel)")
        lines.append("OS: \(osVersion)")
        lines.append("Reported by: \(userEmail.isEmpty ? "unknown" : userEmail) (\(userId))")
        lines.append("Report ID: \(id)")
        return lines.joined(separator: "\n")
    }

    var emailSubject: String {
        "Shifnex \(categoryLabel) — \(String(id.prefix(8)))"
    }
}
